### PR TITLE
[WIP] Direct download D4RL datasets

### PR DESCRIPTION
**Pull request by MateuszGuzek**
## Description

Allow for a direct download of D4RL datasets, without installing D4RL.

## Motivation and Context

TBD, raised during direct discussion with @vmoens 

Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.
You can use the syntax `close #15213` if this solves the issue #15213

- [ ] I have raised an issue to propose this change ([required](https://github.com/pytorch/rl/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [X] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: torchrl/data/datasets/d4rl.py [hunk 1]
```diff
@@ -93,6 +93,188 @@ class D4RLExperienceReplay(TensorDictReplayBuffer):
 
     """
 
+    D4RL_DATASETS = {
```
Comment (vmoens) [inline torchrl/data/datasets/d4rl.py:96]: Let's move this to a separate file, it hampers readability

File: torchrl/data/datasets/d4rl.py [hunk 2]
```diff
@@ -118,19 +300,23 @@ def __init__(
...
+            if not self._has_d4rl:
+                raise ImportError("Could not import d4rl") from self.D4RL_ERR
+
+            if from_env:
+                dataset = self._get_dataset_from_env(name, env_kwargs)
+            else:
+                dataset = self._get_dataset_direct(name, env_kwargs)
         else:
-            dataset = self._get_dataset_direct(name, env_kwargs)
+            dataset = self._get_dataset_direct_download(name, env_kwargs)
```
Comment (vmoens) [inline torchrl/data/datasets/d4rl.py:319]: I guess that `from_env` will be disabled in this case, perhaps we should check that it is not on

**Review by vmoens — COMMENTED** [review]
Good start!

File: torchrl/data/datasets/d4rl.py [hunk 3]
```diff
@@ -149,6 +158,21 @@ def __init__(
...
+    def _get_dataset_direct_download(self, name, env_kwargs):
+        """Directly download and use a D4RL dataset."""
+        if env_kwargs:
+            raise RuntimeError("Cannot pass env_kwargs when `direct_download=True`.")
+        url = D4RL_DATASETS.get(name, None)
+        if url is None:
+            raise KeyError(f"Env {name} not found.")
+        h5path = _download_dataset_from_url(url)
+        # h5path_parent = Path(h5path).parent
+        dataset = PersistentTensorDict.from_h5(h5path)
```
Comment (MateuszGuzek) [inline torchrl/data/datasets/d4rl.py:170]: It fails on the 'metadata' key, in the current D4RL implementation metadata is explicitly removed. So we either need to change peristent.py in tensordict, or write our processing here.

`
        if "metadata" in dataset.keys():
            metadata = dataset.get("metadata")
            dataset = dataset.exclude("metadata")
            self.metadata = metadata
            # find batch size
            dataset = make_tensordict(dataset.flatten_keys("/").to_dict())
            dataset = dataset.unflatten_keys("/")
`

**Pull request closed without merging**